src/Psalm/Internal/Analyzer/Statements/Expression/BinaryOp/ConcatAnalyzer.php

```diff
@@ -277,6 +278,41 @@ public static function analyze(
                     }
                 }
             }
+        } elseif ($left_type || $right_type) {
+            /**
+             * @var Union $known_operand
+             */
+            $known_operand = $right_type ?? $left_type;
+
+            $numeric_type = new Union([
```

Comment:
I think this is a wrong copy/paste because you can't assume a numeric type concatenated with anything will still produce a numeric type.

You can make this assumption with non-empty-string and non-falsy-string but it stops there. So instead of building an Union here, you should start one from scratch with something like Type::getNonEmptyString()